src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -449,6 +449,56 @@ ResultCode GlobalizationNative_GetSortHandle(const char* lpLocaleName, SortHandl
     return GetResultCode(err);
 }
 
+static const char* BreakIteratorRule = "$CR      = [\\p{Grapheme_Cluster_Break = CR}]; \n"          \
+                                       "$LF      = [\\p{Grapheme_Cluster_Break = LF}]; \n"          \
+                                       "$Control = [[\\p{Grapheme_Cluster_Break = Control}]]; \n"   \
+                                       "$Extend  = [[\\p{Grapheme_Cluster_Break = Extend}]];  \n"   \
+                                       "$ZWJ     = [\\p{Grapheme_Cluster_Break = ZWJ}];       \n"   \
+                                       "[^$Control $CR $LF] ($Extend | $ZWJ); \n";
+
+// When doing string search operations using ICU, it is internally using a break iterator which doesn't allow breaking between some characters according to
+// the Grapheme Cluster Boundary Rules specified in http://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules.
+// Unfortunately, not all rules will have the desired behavior we need to get in .NET. For example, the rules don't allow breaking between CR '\r' and LF '\n' characters.
+// When searching for "\n" in a string like "\r\n", will get not found result.
+// We are customizing the break iterator to include only the rules which give the desired behavior. Mainly, we include the GB9 rule http://www.unicode.org/reports/tr29/#GB9
+// which doesn't allow breaking before the nonspace marks.
```

Comment:
What about other rules like GB6 - GB8, which control Hangul searching; or GB12 - GB13, which control regional indicators?

For example, let's say that I have the string `"\U0001F1F5\U0001F1F7\U0001F1FA\U0001F1F8"`. This is the regional indicator for the Puerto Rican flag followed by the regional indicator for the USA flag. If I search for the substring `"\U0001F1F7\U0001F1FA"` (regional indicator for the Russian flag), will this result in a match? The normal break rules ensure that this will not result in a match. Should we have a unit test for this to ensure that we didn't accidentally nuke these break rules?

Reply:
I didn't include such rules for two reasons: 
- It is different than `NLS` behavior. Users usually complain about the behavior change from `NLS.
- The most important reason is we need to differentiate between the string search operation and unit/word/phrase breaking. These are really two different operations and shouldn't be handled differently. If I am parsing a document and trying to find the boundaries of the characters, words, or phrases, then these rules make sense to apply at that time. But if I am searching for a string, these rules don't make sense at that time. Think about as `\r\n` case. If trying to find the boundaries, then not allow breaking between `\r` and `\n` make sense. But if you search `\n` it doesn't make sense not finding it in `\r\n`. Also, ICU has a bug that by default using the break iterator while the documentation is clear telling if not intentionally passing a breaker, then shouldn't break between characters and only the collation operation only should determine the search result. 

in https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/usearch_8h.html#ac0bf8411b693a14ed35d2a8883225cd7 it states:

```
breakiter | A BreakIterator that will be used to restrict the points at which matches are detected. If a match is found, but the match's start or end index is not a boundary as determined by the BreakIterator, the match will be rejected and another will be searched for. If this parameter is NULL, no break detection is attempted
```

Reply:
I want to add, I prefer not adding any more breaking rules from now till we get a real request users want it. Later we can allow our API users to customize the breaking behavior as they want by providing some options doing that.